Technical review: Add WebAuthn Signal API (#37557)
mdn/content: files/en-us/web/api/publickeycredential/signalunknowncredential_static/index.md

- `options`
  - : An object representing the unrecognized credential, which contains the following properties:
    - `credentialId`
      - : A base64url-encoded string representing the [`id` of the credential](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) that was unrecognized.
    - `rpId`
      - : A string representing the [`id` of the RP](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#id_2) that sent the signal.

### Return value

A {{jsxref("Promise")}} that resolves to {{jsxref("undefined")}}.

### Exceptions

The promise rejects with the following exceptions:

- `SecurityError` {{domxref("DOMException")}}
  - : The RP domain is not valid.
- `TypeError` {{domxref("DOMException")}}
  - : The `credentialId` is not a valid base64url-encoded string.

## Description

It is possible for the information stored in a user's authenticator about a [discoverable credential](/en-US/docs/Web/API/Web_Authentication_API#discoverable_credentials_and_conditional_mediation) (for example, [a passkey](https://passkeys.dev/)) to go out sync with the RP server. This usually occurs when the user deletes a credential from the RP site without updating the authenticator.

> The next time they try to sign in with a discoverable credential, the deleted credential will still be presented in the relevant UI, but the attempt to sign in will fail because the RP server won't recognize it. This results in a confusing user experience.

Reviewer: The flow of this isn't quite right - doesn't have to be next time, there could be multiple credentials. The problem is if any credentials are deleted in RP and then selected from UI.

```suggestion
When a user attempts to login, the set of credentials that are presented to the user come from the authenticator.
If they select a credential that has been deleted from the RP server the login will fail, because the RP server won't recognise the returned credential ID.
This is a confusing user experience.
```

Author: This is pretty decent. I've used it in my next commit but with a few tweaks and fixes.

page-type: web-api-static-method
browser-compat: api.PublicKeyCredential.signalUnknownCredential_static
---

{{APIRef("Web Authentication API")}}{{securecontext_header}}

The **`signalUnknownCredential()`** static method of the {{domxref("PublicKeyCredential")}} interface signals to the authenticator that a [credential ID](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) was not recognized by the [relying party](https://en.wikipedia.org/wiki/Relying_party) server.

## Description

It is possible for the information stored in a user's authenticator about a [discoverable credential](/en-US/docs/Web/API/Web_Authentication_API#discoverable_credentials_and_conditional_mediation) (for example, [a passkey](https://passkeys.dev/)) to go out sync with the relying party's server. This usually occurs when the user deletes a credential from the relying party's site without updating the authenticator.

The next time they try to sign in with a discoverable credential, the deleted credential will still be presented in the relevant UI, but the attempt to sign in will fail because the server won't recognize it. This results in a confusing user experience.

> `signalUnknownCredential()` can then be called on the relying party site to tell the authenticator that the credential ID was not recognized. It is up to the authenticator how to handle this information, but the expectation is that it will delete the relevant credential so that the same confusion does not occur again.

Reviewer 2: It would be good to have a discussion on when one would use `signalUnknownCredential` vs `signalAllAcceptedCredentials`.

`signalUnknownCredential` should be used when the user is not authenticated, i.e. after a failed sign in attempt with a passkey that was deleted. That's because it only takes one credential ID. Using `signalAllAcceptedCredentials` for that case would reveal the list of credential IDs for a given user to an unauthenticated party, which may not be desirable.

`signalAllAcceptedCredentials` should be used in all cases when the user is authenticated (after every sign in, and when deleting a credential on the site) -- even if for the latter `signalUnknownCredential` may still work. This is because it sends a complete snapshot of all credentials, so any newly plugged in authenticators have a chance to update all the state.'

(The PR does mention when each should be used, but not why, and it may be worth calling it out).

Author: Hey @nsatragno, thanks so much for the technical review!

To fix this one, I've read carefully through your notes and added a subsection to the "Description" sections of the `signalAllAcceptedCredentials()` and `signalUnknownCredential()` pages to explain this. You'll see this in my next commit.

page-type: web-api-static-method
browser-compat: api.PublicKeyCredential.signalUnknownCredential_static
---

{{APIRef("Web Authentication API")}}{{securecontext_header}}

The **`signalUnknownCredential()`** static method of the {{domxref("PublicKeyCredential")}} interface signals to the authenticator that a [credential ID](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) was not recognized by the [relying party](https://en.wikipedia.org/wiki/Relying_party) server.

## Description

It is possible for the information stored in a user's authenticator about a [discoverable credential](/en-US/docs/Web/API/Web_Authentication_API#discoverable_credentials_and_conditional_mediation) (for example, [a passkey](https://passkeys.dev/)) to go out sync with the relying party's server. This usually occurs when the user deletes a credential from the relying party's site without updating the authenticator.

The next time they try to sign in with a discoverable credential, the deleted credential will still be presented in the relevant UI, but the attempt to sign in will fail because the server won't recognize it. This results in a confusing user experience.

> After this happens, `signalUnknownCredential()` can be called on the relying party site to tell the authenticator that the credential ID was not recognized. It is up to the authenticator how to handle this information, but the expectation is that it will delete the relevant credential so that the same confusion does not occur again.

Reviewer 3: How about adding use cases for this API more explicitly like other two? I'd write:
- When a passkey-based sign-in has failed
- When a passkey is created but fails to register a credential to the server for some reason (I'll add this to our doc later).

Reviewer 3: @nsatragno what do you think of a registration failure use case?

Author: That's a good idea. I've updated it so that it follows the same sentence structure as the others and includes the use cases. See my last commit.

Reviewer 2: The registration failure use case is fine. It somewhat overlaps `signalAllAcceptedCredentials` because if the user can register a passkey, then presumably they're already authenticated.

Author: OK, cool.

Reviewer 3: If a service is designed to register a passkey as part of a sign-up process, a passkey can be created without the user being authenticated.

---
title: "PublicKeyCredential: signalUnknownCredential() static method"
short-title: signalUnknownCredential()
slug: Web/API/PublicKeyCredential/signalUnknownCredential_static
page-type: web-api-static-method
status:
  - experimental
browser-compat: api.PublicKeyCredential.signalUnknownCredential_static
---

{{APIRef("Web Authentication API")}}{{securecontext_header}}{{seecompattable}}

The **`signalUnknownCredential()`** static method of the {{domxref("PublicKeyCredential")}} interface signals to the authenticator that a [credential ID](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) was not recognized by the [relying party](https://en.wikipedia.org/wiki/Relying_party) server.

> ## Description

Reviewer: Goes after syntax, and add explanation of when called.

Author: Done

It is possible for the information stored in a user's authenticator about a [discoverable credential](/en-US/docs/Web/API/Web_Authentication_API#discoverable_credentials_and_conditional_mediation) (for example, [a passkey](https://passkeys.dev/)) to go out sync with the relying party's server. This usually occurs when the user deletes a credential from the relying party's site without updating the authenticator.

> The next time they try to sign in with a discoverable credential, the deleted credential will still be presented in the relevant UI, but the attempt to sign in will fail because the server won't recognize it. This results in a confusing user experience.

Reviewer: "Because the server won't recognize it" - you mean the RP server? (might be worth being precise).

I THINK that what happens here is the browser pops up a bunch of credentials it fetches from the authenticator, and the user selects one, which then gets passed to the relying party. But the relying party has deleted the account or other info that uses that credential. Right?

Author: Yes, this does mean the RP server. I have updated it.

The RP and the authenticator _both_ store pieces of the puzzle. Look at the description on the main Web Authn API landing page in the "Creating a key pair and registering a user" section — it says "The server stores the public key, coupled with the user identity, to remember the credential for future authentications."

When you try to sign in, and the RP is using discoverable credentials, the metadata held by the authenticator is used to populate the sign-in UI with a set of options for the current user, so they can easily choose an account to sign in with.

Unfortunately, if the user does something like delete an account on the RP site, the credentials stored on the RP site and in the authenticator can then go out of sync.

The whole point of these new methods is to allow the RP site to notify the authenticator when something has changed on its side, so it can sync up with it.

> To avoid this issue, `signalUnknownCredential()` should be called on the relying party site each time a discoverable credential-based sign-in fails, or when a discoverable credential is created but fails to be stored on the server for some reason. This tells the authenticator that the credential ID was not recognized. It is up to the authenticator how to handle this information, but the expectation is that it will delete the relevant credential so that the same confusion does not occur again.

Reviewer: > To avoid this issue, `signalUnknownCredential()` should be called on the relying party site each time a discoverable credential-based sign-in fails

Sounds about right

> , or when a discoverable credential is created but fails to be stored on the server for some reason.

I _think_ this is wrong/not quite right. If the credential is not stored then there will be a mismatch between server and authenticator data. That will cause a "discoverable credential-based sign-in failure". As this reads now, it sounds like you can detect the that the server failed to store the credential in some other way and call `signalUnknownCredential()` 

As per https://github.com/mdn/content/pull/37557/files#r1927966758 it is not clear what is stored on the server, so I could be wrong.

Author: I'm honestly not sure. Are you able to help clear this up, @nsatragno? Apologies that I keep invoking you like this, but this API is complex, and has confused a lot of people over the years.

Reviewer 2: Imagine this scenario: the front-end for a website invoked webauthn and got a successful credential.

Now, the front-end wants to store the details (such as the credential id and the public key) on the back-end with a `fetch` call. Unfortunately, the fetch fails for some reason that doesn't leave room to retry.

Now the user is in a bad situation: the next time they want to sign in with an empty allow list, the browser will happily offer the credential they created earlier. However, it won't be recognized by the site. To avoid this situation in the first place, the front-end could call `signalUnknownCredential()` after the failure to store the credential on the server.

Reviewer: >  the front-end could call signalUnknownCredential() _after the failure to store the credential on the server_.

Ah, thanks @nsatragno ! I think you mean "to clean up the credential on the authenticator".

So @chrisdavidmills you are right - this is for a case where the web app created a credential on the authenticator but was unable to store the data in the backend, so instead it removes it from the authenticator. Even though this is correct, we might perhaps modify this slightly as in https://github.com/mdn/content/pull/37557/files#r1931313810

Author: Cool, all sounds good. I've used your suggestion but tweaked it slightly.

### `signalUnknownCredential()` versus `signalAllAcceptedCredentials()`

It may seem like `signalUnknownCredential()` and {{domxref("PublicKeyCredential.signalAllAcceptedCredentials_static", "signalAllAcceptedCredentials()")}} have similar purposes, so what situation should each one be used in?

To be clear, `signalUnknownCredential()` should be used to update the authenticator when the user is not authenticated. It only passes a single `credentialId` to the authenticator — the same one the client just tried to authenticate with. Using `signalAllAcceptedCredentials()` for this purpose would share the entire list of `credentialId`s for a given user with an unauthenticated party, which may not be desirable.

> `signalAllAcceptedCredentials()` should be used in cases where the user is authenticated and you want to update the state of a user's credentials.

Author: I'm not really sure what you are saying I need to change here. You need to document the clarifying bit in the intro and the Description, surely? The Description needs to work as a standalone bit of text, in case someone gets linked straight to it?

> `signalAllAcceptedCredentials()` should be used in cases where the user is authenticated and you want to update the state of a user's credentials.

Reviewer: Easier to show you - done in https://github.com/mdn/content/pull/37557/files#r1931317437

Closing this.

## Syntax

```js-nolint
signalUnknownCredential(options)
```

### Parameters

- `options`
  - : An object representing the unrecognized credential, which contains the following properties:
    - `credentialId`
      - : A base64url-encoded string representing the [`id` of the credential](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) that was unrecognized.
    - `rpId`
      - : A string representing the [`id` of the relying party](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#id_2) that sent the signal.

### Return value

A {{jsxref("Promise")}} that resolves to {{jsxref("undefined")}}.

### Exceptions

> The promise rejects with a `TypeError` {{domxref("DOMException")}} if the `credentialId` is not a valid base64url-encoded string.

Reviewer: For all these there is also a "SecurityError" if the rpid doesn't resolve https://w3c.github.io/webauthn/#sctn-signal-methods-async-rp-id-validation

Author: I added the following line to the "Exceptions" sections:

> It also rejects with a `SecurityError` {{domxref("DOMException")}} if the RP domain is not valid.

I thiiiink that sounds about right?

Reviewer: @chrisdavidmills Yes it does.

Note, not to template here and in others. Should look something like https://github.com/mdn/content/pull/37557/files#r1931308584

## Examples

The following snippet shows typical usage of `signalUnknownCredential()`:

```js
const credential = await navigator.credentials.get({

>   // get() options

Reviewer: So I still don't get the flow so can't tell if this makes sense. Might help if this had some options set here.

It kind of indicates the flow might be that you web app requests a credentials - the browser shows UI and you get get a credential back.

Then you send this as part of the payload to your login endpoint. The server must be storing some kind of ids for credentials it knows about and since it doesn't know this one it returns 404, and so you signal that this particular id is unknown. 

Might be worth explaining this, perhaps showing that the credential you got back has the id you then signal.

What happens in the success case? You got the credential back and passed to your login, and you check against your stored credential ids. But since you got this credential selected by the user, what are you storing on the site? What are the options in get and the returned credential that allow you to know you can trust this credential - i.e. in the server version you have a public key and you get back a signed challenge, but it isn't clear here where the trust comes from.

Author: I don't know the answers here. I pretty much copied these examples verbatim from the Google docs on the subject.

Reviewer: I get this now - its exactly as outlined by Nina in https://github.com/mdn/content/pull/37557#discussion_r1931031816

This does make sense 
- the get is sent without any credential id and a challenge, and is returned with a credential id (as selected by a user) and the challenge signed/certified using the private key associated with the public key associated with that credential ID.
- This gets sent to the RP server in the fetch. The RP attempts to locate the credential id, but it can't find it, so returns a 404. The web app then signals the authenticator of the unknown credential.

As to whether it is good enough, up to you

I personally would have found the following useful:
- Get options showing no credential id - and a note/comment that this is what makes this a discoverable credential.
- Also show the rpid in the option so that we can see it again later matching our signal, and a challenge
- A comment showing the returned credential  id - so that we can see this later being used.
- Perhaps a few words outlining the above - that the server is checking the returned credential ID.

Author: OK, I think I have done all this. Let me know if you think it makes sense!

---
title: "PublicKeyCredential: signalUnknownCredential() static method"
short-title: signalUnknownCredential()
slug: Web/API/PublicKeyCredential/signalUnknownCredential_static
page-type: web-api-static-method
status:
  - experimental
browser-compat: api.PublicKeyCredential.signalUnknownCredential_static
---

{{APIRef("Web Authentication API")}}{{securecontext_header}}{{seecompattable}}

The **`signalUnknownCredential()`** static method of the {{domxref("PublicKeyCredential")}} interface signals to the authenticator that a [credential ID](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) was not recognized by the [relying party](https://en.wikipedia.org/wiki/Relying_party) (RP) server.

> This allows the authenticator to remove credentials that are not allowed by the RP, such as those for deleted accounts, or accounts that were not properly stored on the server. It should only be used when the current user is not authenticated — after sign in fails because the account details were not available to the RP.

Reviewer: It can be used anytime because it doesn't contain any information that exposes info that should only be available to the user.

```suggestion
This allows the authenticator to remove credentials that are not allowed by the RP, such as those for deleted accounts, or accounts that were created and stored on the authenticator but not properly updated on the server. Generally the method is called after sign in fails because the account details were not available to the RP. It can be used even when the current user is not authenticated because it does not expose sensitive information.
```

Author: Like it, updated.

## Syntax

```js-nolint
signalUnknownCredential(options)
```

### Parameters

- `options`
  - : An object representing the unrecognized credential, which contains the following properties:
    - `credentialId`
      - : A base64url-encoded string representing the [`id` of the credential](/en-US/docs/Web/API/PublicKeyCredentialRequestOptions#id) that was unrecognized.
    - `rpId`
      - : A string representing the [`id` of the RP](/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#id_2) that sent the signal.

### Return value

A {{jsxref("Promise")}} that resolves to {{jsxref("undefined")}}.

### Exceptions

> The promise rejects with a `TypeError` {{domxref("DOMException")}} if the `credentialId` is not a valid base64url-encoded string.
> 
> It also rejects with a `SecurityError` {{domxref("DOMException")}} if the RP domain is not valid.

Reviewer: Information good, but not to template, which uses definition lists. This is a slight modification to the template that makes it clear that the exception is a promise rejection, not a throw.

This approach should be used everywhere.

```suggestion
The promise rejects the following exceptions:

 - `TypeError` {{domxref("DOMException")}}
   - : The `credentialId` is not a valid base64url-encoded string.
- `SecurityError` {{domxref("DOMException")}} 
  - : The RP domain is not valid.
```

Author: I knew this, but I was being lazy; thanks for pulling my socks up!

I've updated all three method pages to use this form.

## Description

It is possible for the information stored in a user's authenticator about a [discoverable credential](/en-US/docs/Web/API/Web_Authentication_API#discoverable_credentials_and_conditional_mediation) (for example, [a passkey](https://passkeys.dev/)) to go out sync with the RP server. This usually occurs when the user deletes a credential from the RP site without updating the authenticator.

The next time they try to sign in with a discoverable credential, the deleted credential will still be presented in the relevant UI, but the attempt to sign in will fail because the RP server won't recognize it. This results in a confusing user experience.

> To avoid this issue, `signalUnknownCredential()` should be called on the RP site each time a discoverable credential-based sign-in fails, or when a discoverable credential is created but fails to be stored on the server for some reason. This tells the authenticator that the credential ID was not recognized. It is up to the authenticator how to handle this information, but the expectation is that it will delete the relevant credential so that the same confusion does not occur again.

Reviewer: Perhaps, following https://github.com/mdn/content/pull/37557#discussion_r1931043059

```suggestion
To avoid this issue, `signalUnknownCredential()` should be called on the RP site each time a discoverable credential-based sign-in fails, to inform the authenticator that the credential ID was not recognized. 
It is up to the authenticator how to handle this information, but the expectation is that it will delete the relevant credential so that there is no mismatch in the data stored on the authenticator and relying party.
In addition, `signalUnknownCredential()` might also be called if a web app is able to create a discoverable credential on the authenticator but is, for any reason, unable to upload the credential information to the server.
```

Author: I implemented this one, reads well.

> 
> ### `signalUnknownCredential()` versus `signalAllAcceptedCredentials()`
> 
> It may seem like `signalUnknownCredential()` and {{domxref("PublicKeyCredential.signalAllAcceptedCredentials_static", "signalAllAcceptedCredentials()")}} have similar purposes, so what situation should each one be used in?
> 
> To be clear, `signalUnknownCredential()` should be used to update the authenticator when the user is not authenticated. It only passes a single `credentialId` to the authenticator — the same one the client just tried to authenticate with. Using `signalAllAcceptedCredentials()` for this purpose would share the entire list of `credentialId`s for a given user with an unauthenticated party, which may not be desirable.
> 
> `signalAllAcceptedCredentials()` should be used in cases where the user is authenticated and you want to update the state of a user's credentials.

Reviewer: I would move this overview to the main web api doc with a heading "Authenticator sychronization (signal) methods" or similar. None of this is relevant to signalUnknownCredential because it can be called anywhere.

You might leave this empty, or add a note as I have done

```suggestion

The method can be called even when the current user is not authenticated because it does not expose sensitive information.
```

Author: OK, I've done as you asked. I added the sentence at the bottom of this section to tie that up, and move the whole comparison bit to the main page section. The heading I decided on is "Discoverable credential synchronization methods".

I also removed the comparison similar section from the `signalAllAcceptedCredentials()` page, as I don't think it is needed there either.

The next time they try to sign in with a discoverable credential, the deleted credential will still be presented in the relevant UI, but the attempt to sign in will fail because the server won't recognize it. This results in a confusing user experience.

To avoid this issue, `signalUnknownCredential()` should be called on the relying party site each time a discoverable credential-based sign-in fails, or when a discoverable credential is created but fails to be stored on the server for some reason. This tells the authenticator that the credential ID was not recognized. It is up to the authenticator how to handle this information, but the expectation is that it will delete the relevant credential so that the same confusion does not occur again.

### `signalUnknownCredential()` versus `signalAllAcceptedCredentials()`

It may seem like `signalUnknownCredential()` and {{domxref("PublicKeyCredential.signalAllAcceptedCredentials_static", "PublicKeyCredential.signalAllAcceptedCredentials()")}} have similar purposes, so which one should be used when?

> To be clear, `signalUnknownCredential()` should be used to update the authenticator when authentication _failed_. It only passes a single `credentialId` to the authenticator, minimizing the data shared with an unauthenticated party. Using `signalAllAcceptedCredentials()` for this purpose would share the entire list of `credentialId`s for a given user with an unauthenticated party, which may not be desirable.

Reviewer 2: (roughly same comments apply here)

```suggestion
To be clear, `signalUnknownCredential()` should be used to update the authenticator when the user is not authenticated. It only passes the same `credentialId` the client just tried to the authenticator. Using `signalAllAcceptedCredentials()` for this purpose would share the entire list of `credentialId`s for a given user with an unauthenticated party, which may not be desirable.
```

> `signalAllAcceptedCredentials()` should be used in cases where authentication is _successful_ and you want to update the state of a user's credentials.

Reviewer 2: ```suggestion
`signalAllAcceptedCredentials()` should be used in cases where the user is authenticated and you want to update the state of a user's credentials.
```
